### PR TITLE
First pass implementation

### PR DESCRIPTION
**Pull request by Jake-Shadle**
First pass implementation of our automerge github action.

### Why?

First thing to clarify is why we actually created our own, so here's a short list of other solutions in this space and why they don't (currently) meet our needs.

#### [Github auto-merge](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request)

Github's own auto-merge recently went into beta, and it _almost_ meets our needs, but has a few drawbacks.

* Someone still has to hit the automerge button on the PR, so there's (usually) an additional manual step in the process to actually enable the auto merge
  * This action instead does the reverse, and by default _every_ PR in the repo this action is available in will be automerged by default
  * This action also has 2 ways to opt out on a per PR basis from automerge, namely, making the PR a draft, as well as an optional `block-automerge-label` setting which allows the repo to have a specific label that can be added to a PR to stop it from being automerged for whatever reason, until the label is removed.
* When requiring reviews before a PR can be merged, Github's branch protections have a, frankly, baffling set of criteria, namely around the number of reviewers, which is just a static number that applies to every PR in the repo, and once that number is reached, the PR can be merged, even if there are other reviewers who have been requested who haven't submitted a review!
  * If reviews are required for this action to automerge, it instead requires that there be at least 1 reviewer, but then requires that _all_ requested reviewers have approved a PR before it can be automerged, which in my mind is the much more expected behavior. If a reviewer feels that the set of reviewers needs to be changed, they can change these before submitting their own review to make sure the PR has been seen by all relevant parties.

#### [mergify](https://mergify.io/)

Mergify has the problem that it is an external service that integrates with Github, and we'd rather not use it in internal repos, though it is generally quite good otherwise.

#### Other [automerge](https://github.com/pascalgn/automerge-action) actions

These tend to have similar problems to the [Github auto-merge](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request) since they tend to rely on the same branch protections, or need other actions to actually add/remove labels, and it's easier to just have 1 action that does everything.

#### Note on :crab:

One additional note, is that this action is entirely written in Rust and just exposed as a container image, which (hopefully) means more people on my team would be comfortable making fixes/improvements since we are a Rust house and doing the action in the usual JS/TS was a barrier to entry.

### Options

These are the current options exposed by the action

```yaml
  needs-description-label:
    description: "Label applied to PRs that don't have a description"
    required: false
  required-statuses:
    description: "List of `,` separated statuses the are required to pass for automerge"
    required: true
  ci-passed-label:
    description: "Label applied when all of the required status checks have passed"
    required: false
    default: "s: ci-passed"
  reviewed-label:
    description: "Label applied when all requested reviews have been accepted"
    required: false
  block-merge-label:
    description: "Label that can be applied manually to block automerges until it is removed"
    required: false
  automerge-grace-period:
    description: "Time in milliseconds between when the action detects that the PR is automergeable and when a merge is attempted"
    required: false
  merge-method:
    description: "The merge method to use when merging a PR"
    default: "merge"
    required: false
  github-token:
    description: "The Github token used for all API requests"
    required: true
```

### Missing pieces

One big missing piece that I would like to add would be an `update-method` to allow the action to automatically (when possible) update the PR branch, since branch protections can be placed to require PR branches to be up to date with the base branch before they can be merged, which can be an annoying thing for the PR author to due when everything else is ready but another PR just happened to be merged before them.

### CODE REVIEW

**Review by repi — APPROVED** [review]
looks like a great start!

didn't really review the code, so if you want a more proper review add some additional reviewers for that :)

**Comment by Jake-Shadle** [comment]
Yah I'm not really looking for a code review, more current features/options and anything that is needed before we start using it.

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+name: "octobors"
+description: "Automatically merge"
+inputs:
+  needs-description-label:
+    description: "Label applied to PRs that don't have a description"
+    required: false
+  required-statuses:
+    description: "List of `,` separated statuses the are required to pass for automerge"
```
Comment (lpil) [inline action.yml:8]: Can this be inferred from the branch protection rules? Would be nice to not duplicate this information. Not a big deal though.

File: build_and_push.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+set -eu
+
+cmd="docker"
+
+if ! [[ $(command -v $cmd) ]]; then
+    cmd="podman"
```
Comment (lpil) [inline build_and_push.sh:7]: podman rocks

File: action.yml (hunk 1 above)
Comment (Jake-Shadle) [reply action.yml:8]: This was meant to work in a scenario where you may not want to enable the status check protections in Github.

**Pull request merged**